### PR TITLE
Feature lattice predicates

### PR DESCRIPTION
**Pull request by tonimueller**
Recently, I need to have the monomer ID on the lattice edge for a reaction mechanism. 
For that purpose I wrote the feature FeatureExcludedVolumeIdOnLattice.h which uses a predicate as template parameter. The predicates are specified within the LatticePredicate.h and up to now, there are four functors which can be passed to the feature. 
Of course, the feature has a test :-)

### CODE REVIEW

File: include/LeMonADE/core/MoleculesRead.h [hunk 1]
```diff
@@ -152,6 +153,79 @@ void ReadBonds<IngredientsType>::execute()
...
+};
+/***********************************************************************/
+/**
+ * @brief Executes the reading routine to extract \b !remove_bonds.
+ *
+ * @throw <std::runtime_error> if broken bond could not be read.
+ **/
+
+template < class IngredientsType >
+void ReadBreaks<IngredientsType>::execute()
```
Comment (MartinWenge) [inline include/LeMonADE/core/MoleculesRead.h]: To be more precise and comparable to ReadAddBonds/WriteRemoveBonds I really prefere to rename these classes
```suggestion
void ReadRemoveBonds<IngredientsType>::execute()
```

File: include/LeMonADE/core/MoleculesRead.h [hunk 2]
```diff
@@ -152,6 +153,79 @@ void ReadBonds<IngredientsType>::execute()
...
 
+/***********************************************************************/
+/**
+ * @class ReadBreaks
+ * @brief Handles BFM-File-Reads \b !remove_bonds.
+ *
+ * @tparam IngredientsType Ingredients class storing all system information.
+ **/
+template < class IngredientsType >
+class ReadBreaks: public ReadToDestination < IngredientsType >
```
Comment (MartinWenge) [inline include/LeMonADE/core/MoleculesRead.h]:
```suggestion
class ReadRemoveBonds: public ReadToDestination < IngredientsType >
```

File: include/LeMonADE/core/MoleculesRead.h [hunk 3]
```diff
@@ -152,6 +153,79 @@ void ReadBonds<IngredientsType>::execute()
   std::cout << "ReadBonds:execute() : done " <<std::endl;
 }
 
+/***********************************************************************/
+/**
+ * @class ReadBreaks
```
Comment (MartinWenge) [inline include/LeMonADE/core/MoleculesRead.h]:
```suggestion
 * @class ReadRemoveBonds
```

File: include/LeMonADE/core/MoleculesRead.h [hunk 4]
```diff
@@ -152,6 +153,79 @@ void ReadBonds<IngredientsType>::execute()
...
+ * @brief Handles BFM-File-Reads \b !remove_bonds.
+ *
+ * @tparam IngredientsType Ingredients class storing all system information.
+ **/
+template < class IngredientsType >
+class ReadBreaks: public ReadToDestination < IngredientsType >
+{
+ public:
+	//! Empty Constructor, but delegates call to the Feature
+	ReadBreaks(IngredientsType& destination):ReadToDestination< IngredientsType > (destination){}
```
Comment (MartinWenge) [inline include/LeMonADE/core/MoleculesRead.h]:
```suggestion
	ReadRemoveBonds(IngredientsType& destination):ReadToDestination< IngredientsType > (destination){}
```

File: include/LeMonADE/core/MoleculesRead.h [hunk 5]
```diff
@@ -152,6 +153,79 @@ void ReadBonds<IngredientsType>::execute()
...
+/**
+ * @brief Executes the reading routine to extract \b !remove_bonds.
+ *
+ * @throw <std::runtime_error> if broken bond could not be read.
+ **/
+
+template < class IngredientsType >
+void ReadBreaks<IngredientsType>::execute()
+{
+  std::cout << "ReadBreaks:execute() : updating " <<std::endl;
```
Comment (MartinWenge) [inline include/LeMonADE/core/MoleculesRead.h]:
```suggestion
  std::cout << "ReadRemoveBonds:execute() : updating " <<std::endl;
```

File: include/LeMonADE/core/MoleculesRead.h [hunk 6]
```diff
@@ -152,6 +153,79 @@ void ReadBonds<IngredientsType>::execute()
...
+    //initialize a stringstream with the line for easier processing
+    std::stringstream stream(line);
+
+    //read bond partners
+    stream>>a>>b;
+    //throw exception if something went wrong
+    if(stream.fail()){
+      
+    	std::stringstream errormessage;
+      errormessage<<"ReadBreaks<IngredientsType>::execute()\n"
```
Comment (MartinWenge) [inline include/LeMonADE/core/MoleculesRead.h]:
```suggestion
      errormessage<<"ReadRemoveBonds<IngredientsType>::execute()\n"
```

File: include/LeMonADE/core/MoleculesRead.h [hunk 7]
```diff
@@ -152,6 +153,79 @@ void ReadBonds<IngredientsType>::execute()
...
+    }
+      
+    //if still here, add broken bond to bondset
+    this->getDestination().modifyMolecules().disconnect(a-1,b-1);
+    nBreaks++;
+    //read next line from file
+    getline(this->getInputStream(),line);
+  }
+  
+  std::cout << "ReadBreaks:execute() : done " <<std::endl;
```
Comment (MartinWenge) [inline include/LeMonADE/core/MoleculesRead.h]:
```suggestion
  std::cout << "ReadRemoveBonds:execute() : done " <<std::endl;
```

File: include/LeMonADE/feature/FeatureMoleculesIO.h [hunk 8]
```diff
@@ -129,6 +130,7 @@ void FeatureMoleculesIO::exportRead(FileImport<IngredientsType>& fileReader)
 	fileReader.registerRead("!number_of_monomers", new  ReadNrOfMonomers< IngredientsType > (destination) );
 	fileReader.registerRead("!bonds", new  ReadBonds< IngredientsType > (destination) );
 	fileReader.registerRead("!add_bonds", new  ReadBonds< IngredientsType > (destination) );
+	fileReader.registerRead("!remove_bonds", new  ReadBreaks< IngredientsType > (destination) );
```
Comment (MartinWenge) [inline include/LeMonADE/feature/FeatureMoleculesIO.h]:
```suggestion
	fileReader.registerRead("!remove_bonds", new  ReadRemoveBonds< IngredientsType > (destination) );
```

File: include/LeMonADE/feature/FeatureMoleculesIOUnsaveCheck.h [hunk 9]
```diff
@@ -0,0 +1,174 @@
...
+*/
+
+template < class IngredientsType >
+void FeatureMoleculesIOUnsaveCheck::exportRead(FileImport<IngredientsType>& fileReader)
+{
+	IngredientsType& destination=fileReader.getDestination();
+	fileReader.registerRead("!number_of_monomers", new  ReadNrOfMonomers< IngredientsType > (destination) );
+	fileReader.registerRead("!bonds", new  ReadBonds< IngredientsType > (destination) );
+	fileReader.registerRead("!add_bonds", new  ReadBonds< IngredientsType > (destination) );
+	fileReader.registerRead("!remove_bonds", new  ReadBreaks< IngredientsType > (destination) );
```
Comment (MartinWenge) [inline include/LeMonADE/feature/FeatureMoleculesIOUnsaveCheck.h]:
```suggestion
	fileReader.registerRead("!remove_bonds", new  ReadRemoveBonds< IngredientsType > (destination) );
```

File: include/LeMonADE/utility/LatticePredicates.h [hunk 10]
```diff
@@ -0,0 +1,84 @@
...
+{
+  template <class MoleculesType>
+  uint32_t operator() (const MoleculesType& m, int i){ 
+    uint32_t LatticeEntry(i+1);
+    uint32_t Attribute(m[i].getAttributeTag());
+    LatticeEntry=(LatticeEntry <<6) + Attribute;
+    return LatticeEntry;
+  }
+};
+#endif
```
Comment (MartinWenge) [inline include/LeMonADE/utility/LatticePredicates.h]: adding newline to eof
```suggestion
#endif

```

File: include/LeMonADE/utility/LatticePredicates.h [hunk 11]
```diff
@@ -0,0 +1,84 @@
...
+  template <class MoleculesType>
+  uint32_t operator() (const MoleculesType& m, int i){ return i+1;}
+};
+
+/**
+ * @struct AttributeANDMonomerID
+ *
+ * @brief Functor returning attribute and monomer id .
+ *
+ * @details Thie functor writes on the first two bits the attribute and on the remaining bits the monomer id.
```
Comment (MartinWenge) [inline include/LeMonADE/utility/LatticePredicates.h]: just a typo
```suggestion
 * @details This functor writes on the first two bits the attribute and on the remaining bits the monomer id.
```

File: include/LeMonADE/updater/UpdaterAbstractCreate.h [hunk 12]
```diff
@@ -252,6 +258,206 @@ bool UpdaterAbstractCreate<IngredientsType>::addMonomerInsideConnectedPair(uint3
...
+    //try at most 30 random bondvectors to find a new monomer position
+    for(uint i=0;i<30;i++){
+      // get a random bondvector
+      VectorInt3 bondvector(randomBondvector());
+      // set position of new monomer
+      addmove.setPosition(ingredients.getMolecules()[indexA]+bondvector);
+    
+      // check new position (excluded volume, other features)
+      if(addmove.check(ingredients)==true){
+	//check the new bondvector bewten the new monomer and indexB
```
Comment (MartinWenge) [inline include/LeMonADE/updater/UpdaterAbstractCreate.h]: just a typo
```suggestion
	//check the new bondvector between the new monomer and indexB
```

File: include/LeMonADE/updater/UpdaterAbstractCreate.h [hunk 13]
```diff
@@ -252,6 +258,206 @@ bool UpdaterAbstractCreate<IngredientsType>::addMonomerInsideConnectedPair(uint3
   }
   return false;
 }
+/******************************************************************************/
+/**
+ * @brief function to add a new monomer two already existing ones.
```
Comment (MartinWenge) [inline include/LeMonADE/updater/UpdaterAbstractCreate.h]: Can you add in this comment, that all monomers are supposed to be connected afterwards

**Review by MartinWenge — CHANGES_REQUESTED** [review]
Nice feature, I really like the additions to UpdaterAbstractCreate :+1: 

Please check the renaming of the readBreaks class, I would prefere it.

I passed all the tests and my own tool work with this version. Go for it :o)

File: include/LeMonADE/updater/UpdaterAbstractCreate.h (hunk 13 above)
Comment (tonimueller) [inline include/LeMonADE/updater/UpdaterAbstractCreate.h]:
```suggestion
 * @brief function to add a new monomer to two already existing ones such that all three monomers are connected.
```

Comment (tonimueller) [inline include/LeMonADE/updater/UpdaterAbstractCreate.h]:
```suggestion
 * @brief function to add a new monomer to two already existing ones such that all three are connected afterwards
```

**Review by tonimueller — COMMENTED** [review]
Changed the comment for the function 'addMonomerInsideConnectedPair'

**Comment by tonimueller** [comment]
I accepted all changes and the pull request can be closed I think

**Review by MartinWenge — APPROVED** [review]

**Pull request merged**